sdl_viewer/src/terrain_drawer/layer.rs

```diff
@@ -0,0 +1,232 @@
+use crate::graphic::moving_window_texture::GlMovingWindowTexture;
+use crate::graphic::tiled_texture_loader::TiledTextureLoader;
+use crate::graphic::uniform::GlUniform;
+use crate::graphic::GlProgram;
+use crate::terrain_drawer::read_write::Metadata;
+use cgmath::{Decomposed, Matrix4, Vector2, Vector3, Zero};
+use image::{ImageBuffer, LumaA, Rgba};
+use point_viewer::math::Isometry3;
+use std::convert::TryInto;
+use std::rc::Rc;
+
+const F64_MAX_INT: i64 = 9_007_199_254_740_992;
+
+struct HeightAndColor {
+    height: ImageBuffer<LumaA<f32>, Vec<f32>>,
+    color: ImageBuffer<Rgba<u8>, Vec<u8>>,
+}
+
+pub struct TerrainLayer {
+    grid_coordinates: GridCoordinateFrame,
+    // The terrain pos is the coordinate of the lower corner of the terrain.
+    // It converted to f64 at the latest possible moment, so that all
+    // calculations have 64-bit integer precision.
+    terrain_pos: Vector2<i64>,
+    u_terrain_pos: GlUniform<Vector2<f64>>,
+    height_tiles: TiledTextureLoader<LumaA<f32>>,
+    color_tiles: TiledTextureLoader<Rgba<u8>>,
+    heightmap: GlMovingWindowTexture<LumaA<f32>>,
+    colormap: GlMovingWindowTexture<Rgba<u8>>,
+    // The texture size shouldn't be negative or larger than u32::MAX, but it's
+    // more convenient for our calculations to store it as an i64.
+    texture_size: i64,
+}
+
+#[allow(dead_code)]
+impl TerrainLayer {
+    pub fn new<P: AsRef<std::path::Path>>(
+        program: &GlProgram,
+        path: P,
+        texture_size: u32,
+    ) -> Result<Self, std::io::Error> {
+        println!("Loading terrain.");
```

Comment:
It's almost surely out of scope of this PR, but how about using some lightweight logging library instead of `println!`?

Reply:
I just removed it for now. I'm also not super happy with random prints in the code.